### PR TITLE
Solve the concurrent map write for metaserver handler.go

### LINKED ISSUE #2928: edgecore runs for a long time and crashes.
**What happened**:
```log
Jun 25 12:45:42 dke-edge1 edgecore[2861]: I0625 12:45:42.131231    2861 record.go:19] Warning DNSConfigForming Nameserver limits were exceeded, some nameservers have been omitted, the applied nameserver line is:
Jun 25 12:45:42 dke-edge1 edgecore[2861]: E0625 12:45:42.131251    2861 dns.go:135] Nameserver limits were exceeded, some nameservers have been omitted, the applied nameserver line is: 172.17.0.1 114.114.114.114
Jun 25 12:45:42 dke-edge1 edgecore[2861]: I0625 12:45:42.131336    2861 edged.go:1015] consume added pod [ubuntu] successfully
Jun 25 12:45:42 dke-edge1 edgecore[2861]: I0625 12:45:42.270453    2861 edged_volumes.go:54] Using volume plugin "kubernetes.io/empty-dir" to mount wrapped_default-token-6wktb
Jun 25 12:45:42 dke-edge1 edgecore[2861]: I0625 12:45:42.270550    2861 edged_volumes.go:54] Using volume plugin "kubernetes.io/empty-dir" to mount wrapped_conf
Jun 25 12:45:42 dke-edge1 edgecore[2861]: I0625 12:45:42.270587    2861 edged_volumes.go:54] Using volume plugin "kubernetes.io/empty-dir" to mount wrapped_default-token-9bs5d
Jun 25 12:45:48 dke-edge1 edgecore[2861]: I0625 12:45:48.355302    2861 edged_status.go:198] Sync VolumesInUse: []
Jun 25 12:45:50 dke-edge1 edgecore[2861]: I0625 12:45:50.317731    2861 process.go:525] start to sync pod status in edge-store to cloud
Jun 25 12:45:58 dke-edge1 edgecore[2861]: I0625 12:45:58.391570    2861 edged_status.go:198] Sync VolumesInUse: []
Jun 25 12:46:05 dke-edge1 edgecore[2861]: I0625 12:46:05.512629    2861 storage.go:159] [metaserver/reststorage] successfully apply for a watch listener (/api/v1/secrets) through cloud
Jun 25 12:46:05 dke-edge1 edgecore[2861]: W0625 12:46:05.512706    2861 watcher.go:91] base storage now only support rev == 0, but get rev == 541840, force set to 0!
Jun 25 12:46:05 dke-edge1 edgecore[2861]: I0625 12:46:05.513002    2861 watcher.go:212] start watching, rev:0
Jun 25 12:46:05 dke-edge1 edgecore[2861]: I0625 12:46:05.513090    2861 imitator.go:141] /v1, Resource=secrets,,
Jun 25 12:46:05 dke-edge1 edgecore[2861]: I0625 12:46:05.524908    2861 watcher.go:184] get storage revision:570555
Jun 25 12:46:08 dke-edge1 edgecore[2861]: I0625 12:46:08.130946    2861 edged.go:900] worker [1] get pod addition item [vvd-f86g2]
Jun 25 12:46:08 dke-edge1 edgecore[2861]: I0625 12:46:08.130985    2861 edged.go:968] start to consume added pod [vvd-f86g2]
Jun 25 12:46:08 dke-edge1 edgecore[2861]: I0625 12:46:08.131572    2861 record.go:19] Warning DNSConfigForming Nameserver limits were exceeded, some nameservers have been omitted, the applied nameserver line is:
Jun 25 12:46:08 dke-edge1 edgecore[2861]: E0625 12:46:08.131589    2861 dns.go:135] Nameserver limits were exceeded, some nameservers have been omitted, the applied nameserver line is: 172.17.0.1 114.114.114.114
Jun 25 12:46:08 dke-edge1 edgecore[2861]: I0625 12:46:08.131654    2861 edged.go:1015] consume added pod [vvd-f86g2] successfully
Jun 25 12:46:08 dke-edge1 edgecore[2861]: I0625 12:46:08.249975    2861 edged_volumes.go:54] Using volume plugin "kubernetes.io/empty-dir" to mount wrapped_default-token-9bs5d
Jun 25 12:46:08 dke-edge1 edgecore[2861]: I0625 12:46:08.429886    2861 edged_status.go:198] Sync VolumesInUse: []
Jun 25 12:46:13 dke-edge1 edgecore[2861]: fatal error: concurrent map read and map write
Jun 25 12:46:13 dke-edge1 edgecore[2861]: goroutine 162 [running]:
Jun 25 12:46:13 dke-edge1 edgecore[2861]: runtime.throw(0x386b6d1, 0x21)
Jun 25 12:46:13 dke-edge1 edgecore[2861]: /root/.g/go/src/runtime/panic.go:1116 +0x72 fp=0xc000e9e3d8 sp=0xc000e9e3a8 pc=0x125b562
Jun 25 12:46:13 dke-edge1 edgecore[2861]: runtime.mapaccess2_faststr(0x341d880, 0xc000cf6540, 0x383ac95, 0xf, 0xc002928f48, 0x125ed01)
Jun 25 12:46:13 dke-edge1 edgecore[2861]: /root/.g/go/src/runtime/map_faststr.go:116 +0x47c fp=0xc000e9e448 sp=0xc000e9e3d8 pc=0x1239a6c
Jun 25 12:46:13 dke-edge1 edgecore[2861]: k8s.io/apimachinery/pkg/apis/meta/v1/unstructured.NestedFieldNoCopy(0xc000cf63f0, 0xc000e9e600, 0x2, 0x2, 0x0, 0xc000ab7e90, 0xc001093501, 0x0, 0x0)
Jun 25 12:46:13 dke-edge1 edgecore[2861]: /root/gopath/src/github.com/kubeedge/kubeedge/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go:61 +0xce fp=0xc000e9e4e0 sp=0xc000e9e448 pc=0x1d6bbbe
Jun 25 12:46:13 dke-edge1 edgecore[2861]: k8s.io/apimachinery/pkg/apis/meta/v1/unstructured.NestedString(0xc000cf63f0, 0xc000e9e600, 0x2, 0x2, 0xc000e9e5a0, 0x382b853, 0x8, 0xc000e9e5e0, 0x12310e3)
Jun 25 12:46:13 dke-edge1 edgecore[2861]: /root/gopath/src/github.com/kubeedge/kubeedge/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go:75 +0x67 fp=0xc000e9e578 sp=0xc000e9e4e0 pc=0x1d6be67
Jun 25 12:46:13 dke-edge1 edgecore[2861]: k8s.io/apimachinery/pkg/apis/meta/v1/unstructured.getNestedString(0xc000cf63f0, 0xc000e9e600, 0x2, 0x2, 0x373f560, 0x380ab60)
Jun 25 12:46:13 dke-edge1 edgecore[2861]: /root/gopath/src/github.com/kubeedge/kubeedge/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go:278 +0x4d fp=0xc000e9e5d0 sp=0xc000e9e578 pc=0x1d6dd5d
Jun 25 12:46:13 dke-edge1 edgecore[2861]: k8s.io/apimachinery/pkg/apis/meta/v1/unstructured.(*Unstructured).GetResourceVersion(0xc0011d4e30, 0xc0011d4e30, 0x7f058c0b9a50)
Jun 25 12:46:13 dke-edge1 edgecore[2861]: /root/gopath/src/github.com/kubeedge/kubeedge/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go:282 +0x85 fp=0xc000e9e630 sp=0xc000e9e5d0 pc=0x1d7
Jun 25 12:46:13 dke-edge1 edgecore[2861]: github.com/kubeedge/kubeedge/edge/pkg/metamanager/metaserver/kubernetes/storage/sqlite/imitator/watchhook.Trigger(0x382b853, 0x8, 0x3db2580, 0xc0011d4e30)
Jun 25 12:46:13 dke-edge1 edgecore[2861]: /root/gopath/src/github.com/kubeedge/kubeedge/edge/pkg/metamanager/metaserver/kubernetes/storage/sqlite/imitator/watchhook/hooks.go:67 +0x36d fp=0xc000e9e880 sp=0xc000e9
Jun 25 12:46:13 dke-edge1 edgecore[2861]: github.com/kubeedge/kubeedge/edge/pkg/metamanager/metaserver/kubernetes/storage/sqlite/imitator.(*imitator).Inject(0xc000631500, 0xc00359ff50, 0x24, 0x0, 0x0, 0x17a417d5
Jun 25 12:46:13 dke-edge1 edgecore[2861]: /root/gopath/src/github.com/kubeedge/kubeedge/edge/pkg/metamanager/metaserver/kubernetes/storage/sqlite/imitator/imitator.go:59 +0x271 fp=0xc000e9e950 sp=0xc000e9e880 pc
Jun 25 12:46:13 dke-edge1 edgecore[2861]: github.com/kubeedge/kubeedge/edge/pkg/metamanager.(*metaManager).processUpdate(0xc0006f0620, 0xc00359ff50, 0x24, 0x0, 0x0, 0x17a417d5f7b, 0xc001093060, 0x6, 0x0, 0xc001b
Jun 25 12:46:13 dke-edge1 edgecore[2861]: /root/gopath/src/github.com/kubeedge/kubeedge/edge/pkg/metamanager/process.go:222 +0xd9 fp=0xc000e9fce0 sp=0xc000e9e950 pc=0x2eec299
Jun 25 12:46:13 dke-edge1 edgecore[2861]: github.com/kubeedge/kubeedge/edge/pkg/metamanager.(*metaManager).process(0xc0006f0620, 0xc00359ff50, 0x24, 0x0, 0x0, 0x17a417d5f7b, 0xc001093060, 0x6, 0x0, 0xc001ba0700,
Jun 25 12:46:13 dke-edge1 edgecore[2861]: /root/gopath/src/github.com/kubeedge/kubeedge/edge/pkg/metamanager/process.go:646 +0x1e4 fp=0xc000e9fd90 sp=0xc000e9fce0 pc=0x2ef1e04
Jun 25 12:46:13 dke-edge1 edgecore[2861]: github.com/kubeedge/kubeedge/edge/pkg/metamanager.(*metaManager).runMetaManager.func1(0xc0006f0620)
Jun 25 12:46:13 dke-edge1 edgecore[2861]: /root/gopath/src/github.com/kubeedge/kubeedge/edge/pkg/metamanager/process.go:680 +0x2e0 fp=0xc000e9ffd8 sp=0xc000e9fd90 pc=0x2ef31e0
Jun 25 12:46:13 dke-edge1 edgecore[2861]: runtime.goexit()
Jun 25 12:46:13 dke-edge1 edgecore[2861]: /root/.g/go/src/runtime/asm_amd64.s:1373 +0x1 fp=0xc000e9ffe0 sp=0xc000e9ffd8 pc=0x128f151
Jun 25 12:46:13 dke-edge1 edgecore[2861]: created by github.com/kubeedge/kubeedge/edge/pkg/metamanager.(*metaManager).runMetaManager
Jun 25 12:46:13 dke-edge1 edgecore[2861]: /root/gopath/src/github.com/kubeedge/kubeedge/edge/pkg/metamanager/process.go:670 +0x3f
Jun 25 12:46:13 dke-edge1 edgecore[2861]: goroutine 1 [chan receive, 1517 minutes]:
Jun 25 12:46:13 dke-edge1 edgecore[2861]: github.com/kubeedge/beehive/pkg/core.GracefulShutdown()
Jun 25 12:46:13 dke-edge1 edgecore[2861]: /root/gopath/src/github.com/kubeedge/kubeedge/vendor/github.com/kubeedge/beehive/pkg/core/core.go:33 +0x16d
Jun 25 12:46:13 dke-edge1 edgecore[2861]: github.com/kubeedge/beehive/pkg/core.Run()
Jun 25 12:46:13 dke-edge1 edgecore[2861]: /root/gopath/src/github.com/kubeedge/kubeedge/vendor/github.com/kubeedge/beehive/pkg/core/core.go:50 +0x25
Jun 25 12:46:13 dke-edge1 edgecore[2861]: github.com/kubeedge/kubeedge/edge/cmd/edgecore/app.NewEdgeCoreCommand.func1(0xc0003d38c0, 0x562df70, 0x0, 0x0)
Jun 25 12:46:13 dke-edge1 edgecore[2861]: /root/gopath/src/github.com/kubeedge/kubeedge/edge/cmd/edgecore/app/server.go:99 +0x3d2
Jun 25 12:46:13 dke-edge1 edgecore[2861]: github.com/spf13/cobra.(*Command).execute(0xc0003d38c0, 0xc00004e1b0, 0x0, 0x0, 0xc0003d38c0, 0xc00004e1b0)
Jun 25 12:46:13 dke-edge1 edgecore[2861]: /root/gopath/src/github.com/kubeedge/kubeedge/vendor/github.com/spf13/cobra/command.go:846 +0x29d
Jun 25 12:46:13 dke-edge1 edgecore[2861]: github.com/spf13/cobra.(*Command).ExecuteC(0xc0003d38c0, 0xc000098058, 0x0, 0x0)
Jun 25 12:46:13 dke-edge1 edgecore[2861]: /root/gopath/src/github.com/kubeedge/kubeedge/vendor/github.com/spf13/cobra/command.go:950 +0x349
Jun 25 12:46:13 dke-edge1 edgecore[2861]: github.com/spf13/cobra.(*Command).Execute(...)
Jun 25 12:46:13 dke-edge1 edgecore[2861]: /root/gopath/src/github.com/kubeedge/kubeedge/vendor/github.com/spf13/cobra/command.go:887
Jun 25 12:46:13 dke-edge1 edgecore[2861]: main.main()
Jun 25 12:46:13 dke-edge1 edgecore[2861]: /root/gopath/src/github.com/kubeedge/kubeedge/edge/cmd/edgecore/edgecore.go:16 +0x64
Jun 25 12:46:13 dke-edge1 edgecore[2861]: goroutine 6 [chan receive]:
Jun 25 12:46:13 dke-edge1 edgecore[2861]: k8s.io/klog/v2.(*loggingT).flushDaemon(0x5601380)
Jun 25 12:46:13 dke-edge1 edgecore[2861]: /root/gopath/src/github.com/kubeedge/kubeedge/vendor/k8s.io/klog/v2/klog.go:1131 +0x8b
Jun 25 12:46:13 dke-edge1 edgecore[2861]: created by k8s.io/klog/v2.init.0
Jun 25 12:46:13 dke-edge1 edgecore[2861]: /root/gopath/src/github.com/kubeedge/kubeedge/vendor/k8s.io/klog/v2/klog.go:416 +0xd6
Jun 25 12:46:13 dke-edge1 edgecore[2861]: goroutine 57 [select]:
Jun 25 12:46:13 dke-edge1 edgecore[2861]: k8s.io/apimachinery/pkg/apis/meta/v1/unstructured.NestedFieldNoCopy(0xc000cf63f0, 0xc000e9e600, 0x2, 0x2, 0x0, 0xc000ab7e90, 0xc001093501, 0x0, 0x0)
Jun 25 12:46:13 dke-edge1 edgecore[2861]: /root/gopath/src/github.com/kubeedge/kubeedge/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go:61 +0xce fp=0xc000e9e4e0 sp=0xc000e9e448 pc=0x1d6bbbe
Jun 25 12:46:13 dke-edge1 edgecore[2861]: k8s.io/apimachinery/pkg/apis/meta/v1/unstructured.NestedString(0xc000cf63f0, 0xc000e9e600, 0x2, 0x2, 0xc000e9e5a0, 0x382b853, 0x8, 0xc000e9e5e0, 0x12310e3)
Jun 25 12:46:13 dke-edge1 edgecore[2861]: /root/gopath/src/github.com/kubeedge/kubeedge/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go:75 +0x67 fp=0xc000e9e578 sp=0xc000e9e4e0 pc=0x1d6be67
Jun 25 12:46:13 dke-edge1 edgecore[2861]: k8s.io/apimachinery/pkg/apis/meta/v1/unstructured.getNestedString(0xc000cf63f0, 0xc000e9e600, 0x2, 0x2, 0x373f560, 0x380ab60)
Jun 25 12:46:13 dke-edge1 edgecore[2861]: /root/gopath/src/github.com/kubeedge/kubeedge/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go:278 +0x4d fp=0xc000e9e5d0 sp=0xc000e9e578 pc=0x1d6dd5d
Jun 25 12:46:13 dke-edge1 edgecore[2861]: k8s.io/apimachinery/pkg/apis/meta/v1/unstructured.(*Unstructured).GetResourceVersion(0xc0011d4e30, 0xc0011d4e30, 0x7f058c0b9a50)
Jun 25 12:46:13 dke-edge1 edgecore[2861]: /root/gopath/src/github.com/kubeedge/kubeedge/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go:282 +0x85 fp=0xc000e9e630 sp=0xc000e9e5d0 pc=0x1d7
Jun 25 12:46:13 dke-edge1 edgecore[2861]: github.com/kubeedge/kubeedge/edge/pkg/metamanager/metaserver/kubernetes/storage/sqlite/imitator/watchhook.Trigger(0x382b853, 0x8, 0x3db2580, 0xc0011d4e30)
Jun 25 12:46:13 dke-edge1 edgecore[2861]: /root/gopath/src/github.com/kubeedge/kubeedge/edge/pkg/metamanager/metaserver/kubernetes/storage/sqlite/imitator/watchhook/hooks.go:67 +0x36d fp=0xc000e9e880 sp=0xc000e9
Jun 25 12:46:13 dke-edge1 edgecore[2861]: github.com/kubeedge/kubeedge/edge/pkg/metamanager/metaserver/kubernetes/storage/sqlite/imitator.(*imitator).Inject(0xc000631500, 0xc00359ff50, 0x24, 0x0, 0x0, 0x17a417d5
Jun 25 12:46:13 dke-edge1 edgecore[2861]: /root/gopath/src/github.com/kubeedge/kubeedge/edge/pkg/metamanager/metaserver/kubernetes/storage/sqlite/imitator/imitator.go:59 +0x271 fp=0xc000e9e950 sp=0xc000e9e880 pc
Jun 25 12:46:13 dke-edge1 edgecore[2861]: github.com/kubeedge/kubeedge/edge/pkg/metamanager.(*metaManager).processUpdate(0xc0006f0620, 0xc00359ff50, 0x24, 0x0, 0x0, 0x17a417d5f7b, 0xc001093060, 0x6, 0x0, 0xc001b
Jun 25 12:46:13 dke-edge1 edgecore[2861]: /root/gopath/src/github.com/kubeedge/kubeedge/edge/pkg/metamanager/process.go:222 +0xd9 fp=0xc000e9fce0 sp=0xc000e9e950 pc=0x2eec299
Jun 25 12:46:13 dke-edge1 edgecore[2861]: github.com/kubeedge/kubeedge/edge/pkg/metamanager.(*metaManager).process(0xc0006f0620, 0xc00359ff50, 0x24, 0x0, 0x0, 0x17a417d5f7b, 0xc001093060, 0x6, 0x0, 0xc001ba0700,
Jun 25 12:46:13 dke-edge1 edgecore[2861]: /root/gopath/src/github.com/kubeedge/kubeedge/edge/pkg/metamanager/process.go:646 +0x1e4 fp=0xc000e9fd90 sp=0xc000e9fce0 pc=0x2ef1e04
Jun 25 12:46:13 dke-edge1 edgecore[2861]: github.com/kubeedge/kubeedge/edge/pkg/metamanager.(*metaManager).runMetaManager.func1(0xc0006f0620)
Jun 25 12:46:13 dke-edge1 edgecore[2861]: /root/gopath/src/github.com/kubeedge/kubeedge/edge/pkg/metamanager/process.go:680 +0x2e0 fp=0xc000e9ffd8 sp=0xc000e9fd90 pc=0x2ef31e0
Jun 25 12:46:13 dke-edge1 edgecore[2861]: runtime.goexit()
Jun 25 12:46:13 dke-edge1 edgecore[2861]: /root/.g/go/src/runtime/asm_amd64.s:1373 +0x1 fp=0xc000e9ffe0 sp=0xc000e9ffd8 pc=0x128f151
Jun 25 12:46:13 dke-edge1 edgecore[2861]: created by github.com/kubeedge/kubeedge/edge/pkg/metamanager.(*metaManager).runMetaManager
Jun 25 12:46:13 dke-edge1 edgecore[2861]: /root/gopath/src/github.com/kubeedge/kubeedge/edge/pkg/metamanager/process.go:670 +0x3f
Jun 25 12:46:13 dke-edge1 edgecore[2861]: goroutine 1 [chan receive, 1517 minutes]:
Jun 25 12:46:13 dke-edge1 edgecore[2861]: github.com/kubeedge/beehive/pkg/core.GracefulShutdown()
Jun 25 12:46:13 dke-edge1 edgecore[2861]: /root/gopath/src/github.com/kubeedge/kubeedge/vendor/github.com/kubeedge/beehive/pkg/core/core.go:33 +0x16d
Jun 25 12:46:13 dke-edge1 edgecore[2861]: github.com/kubeedge/beehive/pkg/core.Run()
Jun 25 12:46:13 dke-edge1 edgecore[2861]: /root/gopath/src/github.com/kubeedge/kubeedge/vendor/github.com/kubeedge/beehive/pkg/core/core.go:50 +0x25
Jun 25 12:46:13 dke-edge1 edgecore[2861]: github.com/kubeedge/kubeedge/edge/cmd/edgecore/app.NewEdgeCoreCommand.func1(0xc0003d38c0, 0x562df70, 0x0, 0x0)
Jun 25 12:46:13 dke-edge1 edgecore[2861]: /root/gopath/src/github.com/kubeedge/kubeedge/edge/cmd/edgecore/app/server.go:99 +0x3d2
Jun 25 12:46:13 dke-edge1 edgecore[2861]: github.com/spf13/cobra.(*Command).execute(0xc0003d38c0, 0xc00004e1b0, 0x0, 0x0, 0xc0003d38c0, 0xc00004e1b0)
Jun 25 12:46:13 dke-edge1 edgecore[2861]: /root/gopath/src/github.com/kubeedge/kubeedge/vendor/github.com/spf13/cobra/command.go:846 +0x29d
Jun 25 12:46:13 dke-edge1 edgecore[2861]: github.com/spf13/cobra.(*Command).ExecuteC(0xc0003d38c0, 0xc000098058, 0x0, 0x0)
Jun 25 12:46:13 dke-edge1 edgecore[2861]: /root/gopath/src/github.com/kubeedge/kubeedge/vendor/github.com/spf13/cobra/command.go:950 +0x349
Jun 25 12:46:13 dke-edge1 edgecore[2861]: github.com/spf13/cobra.(*Command).Execute(...)
Jun 25 12:46:13 dke-edge1 edgecore[2861]: /root/gopath/src/github.com/kubeedge/kubeedge/vendor/github.com/spf13/cobra/command.go:887
Jun 25 12:46:13 dke-edge1 edgecore[2861]: main.main()
Jun 25 12:46:13 dke-edge1 edgecore[2861]: /root/gopath/src/github.com/kubeedge/kubeedge/edge/cmd/edgecore/edgecore.go:16 +0x64
Jun 25 12:46:13 dke-edge1 edgecore[2861]: goroutine 6 [chan receive]:
Jun 25 12:46:13 dke-edge1 edgecore[2861]: k8s.io/klog/v2.(*loggingT).flushDaemon(0x5601380)
Jun 25 12:46:13 dke-edge1 edgecore[2861]: /root/gopath/src/github.com/kubeedge/kubeedge/vendor/k8s.io/klog/v2/klog.go:1131 +0x8b
Jun 25 12:46:13 dke-edge1 edgecore[2861]: created by k8s.io/klog/v2.init.0
Jun 25 12:46:13 dke-edge1 edgecore[2861]: /root/gopath/src/github.com/kubeedge/kubeedge/vendor/k8s.io/klog/v2/klog.go:416 +0xd6
Jun 25 12:46:13 dke-edge1 edgecore[2861]: goroutine 57 [select]:
Jun 25 12:46:13 dke-edge1 edgecore[2861]: github.com/eclipse/paho%2emqtt%2egolang.keepalive(0xc000762780)
Jun 25 12:46:13 dke-edge1 edgecore[2861]: /root/gopath/src/github.com/kubeedge/kubeedge/vendor/github.com/eclipse/paho.mqtt.golang/ping.go:41 +0x1ea
Jun 25 12:46:13 dke-edge1 edgecore[2861]: created by github.com/eclipse/paho%2emqtt%2egolang.(*client).Connect.func1
Jun 25 12:46:13 dke-edge1 edgecore[2861]: /root/gopath/src/github.com/kubeedge/kubeedge/vendor/github.com/eclipse/paho.mqtt.golang/client.go:306 +0x8a7
Jun 25 12:46:13 dke-edge1 edgecore[2861]: goroutine 94 [select]:
Jun 25 12:46:13 dke-edge1 edgecore[2861]: google.golang.org/grpc/internal/transport.(*controlBuffer).get(0xc00063a2d0, 0x1, 0x0, 0x0, 0x0, 0x0)
Jun 25 12:46:13 dke-edge1 edgecore[2861]: /root/gopath/src/github.com/kubeedge/kubeedge/vendor/google.golang.org/grpc/internal/transport/controlbuf.go:395 +0x122
Jun 25 12:46:13 dke-edge1 edgecore[2861]: google.golang.org/grpc/internal/transport.(*loopyWriter).run(0xc00048e1e0, 0x0, 0x0)

```
**What you expected to happen**:
The tests are all OK.
**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- KubeEdge version(e.g. `cloudcore --version` and `edgecore --version`):
- <details><summary>Cloud nodes Environment:</summary>

  - Hardware configuration (e.g. `lscpu`):
  - OS (e.g. `cat /etc/os-release`):
  - Kernel (e.g. `uname -a`):
  - Go version (e.g. `go version`):
  - Others:

  </details>
- <details><summary>Edge nodes Environment:</summary>

  - edgecore version (e.g. `edgecore --version`):
  - Hardware configuration (e.g. `lscpu`):
  - OS (e.g. `cat /etc/os-release`):
  - Kernel (e.g. `uname -a`):
  - Go version (e.g. `go version`):
  - Others:

  </details>

### PR DESCRIPTION
**Pull request by yz271544**
edge/pkg/metamanager/metaserver/handlerfactory/handler.go
**What type of PR is this?**

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #2928

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

**Review by GsssC — COMMENTED** [review]
/lgtm 
thx for ur contribution

**Review by fisherxu — COMMENTED** [review]
Thanks for the contributions, can we use some like `defer` to unlock? cc @Iceber

File: edge/pkg/metamanager/metaserver/handlerfactory/handler.go [hunk 1]
```diff
@@ -43,25 +45,36 @@ func NewFactory() Factory {
...
+	f.lock.RLock()
 	if h, ok := f.handlers["get"]; ok {
+		f.lock.RUnlock()
 		return h
 	}
+	f.lock.RUnlock()
+	f.lock.Lock()
 	h := handlers.GetResource(f.storage, f.storage, f.scope)
 	f.handlers["get"] = h
+	f.lock.Unlock()
```
Comment (Iceber) [inline edge/pkg/metamanager/metaserver/handlerfactory/handler.go:63]: Too many locks and unlocks can seem confusing.
Whether the read operation can be placed in a separate function.
eg.
```go

func (f *Factory) getHandler(key string) (http.Handler, bool) {
        f.lock.RLock()
        defer f.lock.RUnlock()

        h, ok := f.handlers[key]
        return h, ok
}

func (f *Factory) Get() http.Handler {
        if h, ok := f.getHandler("get"); ok {
                return h
        }

        f.lock.Lock()
        defer f.lock.Lock()
        h := handlers.GetResource(f.storage, f.storage, f.scope)
        f.handlers["get"] = h
        return h
}
```

**Comment by Iceber** [comment]
In addition, it looks like `factory.handler` can be initialized directly in `NewFactory` :thinking:

**Comment by fisherxu** [comment]
ping @yz271544 Ptal for the comments.

File: edge/pkg/metamanager/metaserver/handlerfactory/handler.go [hunk 2]
```diff
@@ -43,25 +45,36 @@ func NewFactory() Factory {
 		scope:             scope.NewRequestScope(),
 		MinRequestTimeout: 1800 * time.Second,
 		handlers:          make(map[string]http.Handler),
+		lock:              &sync.RWMutex{},
```
[the remark this answers is not in the document]
Comment (yz271544) [reply edge/pkg/metamanager/metaserver/handlerfactory/handler.go]: Sorry, Came late, it has been changed.

File: edge/pkg/metamanager/metaserver/handlerfactory/handler.go (hunk 1 above)
Comment (yz271544) [reply edge/pkg/metamanager/metaserver/handlerfactory/handler.go]: Yes, after extracting it, it seems that there is less unnecessary confusion.

**Comment by yz271544** [comment]
Sorry, I was late, and as suggested, I removed the pointer bound to sync.RWMutex.  Extract the handler function, it seems that there is less unnecessary confusion.

File: edge/pkg/metamanager/metaserver/handlerfactory/handler.go [hunk 3]
```diff
@@ -33,33 +34,39 @@ type Factory struct {
...
+	f.lock.Lock()
+	defer f.lock.Unlock()
 	h := handlers.GetResource(f.storage, f.storage, f.scope)
 	f.handlers["get"] = h
 	return h
 }
 
 func (f *Factory) List() http.Handler {
-	if h, ok := f.handlers["list"]; ok {
+	if h, ok := f.getHandler("list"); ok {
```
Comment (muxuelan) [inline edge/pkg/metamanager/metaserver/handlerfactory/handler.go:65]: 想请教一下，为什么cretae、update和patch方法不需要加锁呀？

Comment (Iceber) [reply edge/pkg/metamanager/metaserver/handlerfactory/handler.go:65]: 看起来这个锁是针对 `f.handler`，只有 List 和 Delete 方法使用了 handler 来缓存，不过我有点觉得没有太大的缓存必要，而且并没有带来性能很大的提升(缓存闭包函数)，暂时没看出来这个缓存有什么必要用处 :thinking:

Comment (muxuelan) [reply edge/pkg/metamanager/metaserver/handlerfactory/handler.go:65]: 我根据作者的方法改了下后，在多边端节点的集群中，还是会产生这个fatal
1c567d448d7f3274bcaa913fda308af

e9972d61d0369cfbd8b70a23455ac48

Comment (Iceber) [reply edge/pkg/metamanager/metaserver/handlerfactory/handler.go:65]: 错误栈的其他信息好像和 https://github.com/kubeedge/kubeedge/issues/2928 不太一样，panic 的信息还是 map 并发读写的问题吗

Comment (muxuelan) [reply edge/pkg/metamanager/metaserver/handlerfactory/handler.go:65]: 是的是的，还是map的读写
I2021-08-12 14:49:59.808022   12023 edged.go:1021] consume added pod [edgemesh-gateway-54bf664d67-s487z] successfully
fatal error: concurrent map read and map write

goroutine 1303 [running]:
runtime.throw(0x386f6b0, 0x21)
        /usr/local/go/src/runtime/panic.go:1116 +0x72 fp=0xc002ce74f0 sp=0xc002ce74c0 pc=0x1261362
runtime.mapaccess2(0x341fbe0, 0xc003a168a0, 0xc0037d2bc0, 0xc0037d2bc0, 0xc003396033)
        /usr/local/go/src/runtime/map.go:469 +0x258 fp=0xc002ce7530 sp=0xc002ce74f0 pc=0x123add8
reflect.mapaccess(0x341fbe0, 0xc003a168a0, 0xc0037d2bc0, 0x384f7ab)
        /usr/local/go/src/runtime/map.go:1309 +0x3f fp=0xc002ce7568 sp=0xc002ce7530 pc=0x123cfef
reflect.Value.MapIndex(0x341fbe0, 0xc003a168a0, 0x15, 0x333e8a0, 0xc0037d2bc0, 0x98, 0x37de1e0, 0x341fbe0, 0x341fbe0)
        /usr/local/go/src/reflect/value.go:1166 +0x16d fp=0xc002ce75e0 sp=0xc002ce7568 pc=0x12fbb9d
encoding/json.mapEncoder.encode(0x39596e8, 0xc00377e000, 0x341fbe0, 0xc003a168a0, 0x15, 0x3410100)
        /usr/local/go/src/encoding/json/encode.go:800 +0x305 fp=0xc002ce7758 sp=0xc002ce75e0 pc=0x1595ae5
encoding/json.mapEncoder.encode-fm(0xc00377e000, 0x341fbe0, 0xc003a168a0, 0x15, 0xc003a10100)
        /usr/local/go/src/encoding/json/encode.go:776 +0x64 fp=0xc002ce7798 sp=0xc002ce7758 pc=0x15a1a34
encoding/json.(*encodeState).reflectValue(0xc00377e000, 0x341fbe0, 0xc003a168a0, 0x15, 0xc003a10100)

Comment (Iceber) [reply edge/pkg/metamanager/metaserver/handlerfactory/handler.go:65]: 感觉不太像是pr修改的地方报的错，可以加一些日志，来判断一下是不是`f.List/Delete` 方法导致的

**Review by fisherxu — COMMENTED** [review]
/lgtm
/approve
Merge the pr first, please open issue or pr if have other questions.

File: edge/pkg/metamanager/metaserver/handlerfactory/handler.go (hunk 3 above)
Comment (gy95) [reply edge/pkg/metamanager/metaserver/handlerfactory/handler.go:65]: another stack of concurrent map read/write, branch : release 1.7 .seems a little difficult to find the reason 😞 
https://github.com/kubeedge/kubeedge/issues/3122#issuecomment-912897362

Comment (first-sight12) [reply edge/pkg/metamanager/metaserver/handlerfactory/handler.go:65]: I modified the above, but I found that there was still a problem "fatal error: concurrent map read and map write", and edgecore was still restarted


This is the log of cloud and two edge nodes

[log.tar.gz](https://github.com/kubeedge/kubeedge/files/7109331/log.tar.gz)

Comment (gy95) [reply edge/pkg/metamanager/metaserver/handlerfactory/handler.go:65]: No, this uploaded file dont contain anything about `concurrent map read and map write` , but contain the below info.
 `F0903 17:07:30.901271   32562 server.go:43] listen tcp 127.0.0.1:10350: bind: address already in use`
this is due to address conflict.
Do you upload the exact log file? 😄

Comment (first-sight12) [reply edge/pkg/metamanager/metaserver/handlerfactory/handler.go:65]: This is the log after 2 p.m
[log.tar.gz](https://github.com/kubeedge/kubeedge/files/7109371/log.tar.gz)

Comment (gy95) [reply edge/pkg/metamanager/metaserver/handlerfactory/handler.go:65]: not quite sure about the reason , by the way, Do these logs also contain the logs before you modified the PR modifications?

Comment (first-sight12) [reply edge/pkg/metamanager/metaserver/handlerfactory/handler.go:65]: These logs have not changed after I modified pr. later, I modified it to the original state

**Pull request merged**